### PR TITLE
feat: bas remote functionality

### PR DESCRIPTION
**Pull request by alex-gilin**

### CODE REVIEW

File: packages/app-studio-toolkit/src/bas-remote/devspace/edit.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  getDevSpacesSpec,
+} from "./spec_and_extensions";
+import { DevSpaceInfo, getDevSpaces, PackName } from "./devspace";
+// import { waccess, wOP } from "../utils";
+import { messages } from "../messages";
+import { devspace } from "@sap/bas-sdk";
+import { $enum } from "ts-enum-util";
+import { getJwt } from "../../auth/authentication";
+
+// interface DevSpaceEdit {
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/devspace/edit.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/devspace/connect.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+async function createTunnel(
+  devSpace: DevSpaceNode,
+  progress: Progress<{ message?: string; increment?: number }>
+): Promise<any> {
+  progress.report({ message: "Closing old VPN tunnel to dev-space" });
+  closeTunnel();
+
+  // Start chisel tunnel
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/devspace/connect.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/devspace/connect.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  devSpace: DevSpaceNode
+): Promise<void> {
+  const hostName = await createTunnelAndGetHostName(devSpace);
+  void commands.executeCommand("opensshremotes.openEmptyWindow", {
+    host: hostName,
+  });
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars -- temp suppress
+export async function cmdDevSpaceConnectSameWindow(
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/devspace/connect.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/devspace/devspace.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { window } from "vscode";
+import { getLogger } from "../../logger/logger";
+import { messages } from "../messages";
+import { devspace } from "@sap/bas-sdk";
+import { getJwt } from "../../auth/authentication";
+import { $enum } from "ts-enum-util";
+// import { waccess, wOP } from "../utils";
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/devspace/devspace.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/devspace/devspace.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+export enum DevSpaceStatus {
+  RUNNING = "RUNNING",
+  STARTING = "STARTING",
+  STOPPED = "STOPPED",
+  STOPPING = "STOPPING",
+  ERROR = "ERROR",
+  SAFE_MODE = "SAFE_MODE",
+}
+
+export enum PackName {
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/devspace/devspace.ts]: consider to move in next release

File: packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+function getSshConfigFolderPath(): string {
+  return path.parse(getSshConfigFilePath()).dir;
+}
+
+export async function getPK(
+  adminUrl: string,
+  landscapeUrl: string
+): Promise<string> {
+  return Promise.resolve("");
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    const message = messages.err_config_update(
+      sshConfigFile,
+      (err as Error).toString()
+    );
+    getLogger().error(message);
+    void window.showErrorMessage(message);
+  }
+}
+
+// function getDevChannelPath(): string {
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+  // const channelOsPath: string = getDevChannelPath();
+  const port = 443;
+  const jwt = await getJwt(opt.landscape);
+  if (isEmpty(jwt)) {
+    return;
+  }
+
+  getLogger().info(`Start dev-channel client for ${opt.host} on port ${port}`);
+
+  // void ssh(url.format(urlObj), port, "user", jwt!, { Reconnect: "true" });
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/tunnel/ssh-utils.ts]: outdated

File: packages/app-studio-toolkit/src/bas-remote/utils.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import { isEmpty } from "lodash";
+import axios, { AxiosResponse, RawAxiosRequestHeaders } from "axios";
+import { getHeaders } from "../auth/authentication";
+import { URL } from "node:url";
+
+// let timerStarted : number;
+// let timerInterval : RefreshRate;
+// let timerEnded : number;
+
+export enum RefreshRate {
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/utils.ts]: removed

File: packages/app-studio-toolkit/src/bas-remote/utils.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        }
+        refreshedTime += refreshRate;
+      })
+      .catch((e) => {
+        getLogger().error(`getLandscapes error: ${e.toString()}`);
+      });
+  }, refreshRate);
+}
+
+// export function makeUrl(landscape: string, pathName: string, fragment?: string): string {
```
[the remark this answers is not in the document]
Comment (alex-gilin) [reply packages/app-studio-toolkit/src/bas-remote/utils.ts]: outdated

File: packages/app-studio-toolkit/src/devspace-manager/devspace/devspace.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import { window } from "vscode";
+import { getLogger } from "../../logger/logger";
+import { messages } from "../common/messages";
+import { devspace } from "@sap/bas-sdk";
+import { getJwt } from "../../authentication/auth-utils";
+import { $enum } from "ts-enum-util";
+
+export enum DevSpaceStatus {
```
Comment (idoprz) [inline packages/app-studio-toolkit/src/devspace-manager/devspace/devspace.ts:8]: this should come from the SDK

File: packages/app-studio-toolkit/src/devspace-manager/tunnel/ssh-utils.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+import { homedir } from "os";
+import { DevSpaceNode } from "../tree/treeItems";
+import { getJwt } from "../../authentication/auth-utils";
+import { remotessh } from "@sap/bas-sdk";
+import { ssh } from "./ssh";
+import { URL } from "node:url";
+import { assign, unset } from "lodash";
+const sshConfig = require("ssh-config");
+
+export const SSHD_SOCKET_PORT = 33765;
```
Comment (idoprz) [inline packages/app-studio-toolkit/src/devspace-manager/tunnel/ssh-utils.ts:14]: ports and common urls should come from the SDK

**Review by idoprz — APPROVED** [review]

**Pull request merged**